### PR TITLE
fix(component-testing): Fix socket hangup error

### PR DESCRIPTION
**Pull request by dmtrKovalenko**
This PR fixes the socket hangup error.

## Error description 

This error is coming not from some library, but from nodejs itself. It is possible to see more information by running the process with `NODE_DEBUG=net yarn cy:open`. Socket word here means not a **websocket** but the unix IPC socket.

Webpack HMR works this way: 
1. Create and run socket server together with dev-server
2. Bundles are modified so they are connecting to the socket server and listening for updates <- here problem appears 

Because we running specs in an iframe, and iframes changes too quickly `sockets` connections are not disconnecting properly and we got a system error. 

## The fix

1. Add the `inline: false` option to the wepback-dev-server – it disables webpack's default socket client to be injected inside the bundle
2. Manually connect and listen to the webpack events inside the `runner-ct`. We don't really need to "replace modules" so we basically rerun specs once the new bundle is ready. 

## Potential problems

HMR API is not public, so it can easily change between versions. But it looks like we don't have a choice ¯\_(ツ)_/¯

## Result

It works exactly like the native hmr client, logs all the errors and warnings, except it is not trying to replace the code on the flight and only returning specs. Seems like exactly what we need.

image

### CODE REVIEW

File: packages/runner-ct/src/lib/state.ts [hunk 1]
```diff
@@ -194,7 +196,20 @@ export default class State {
   }
 
   @action setSpec (spec: Cypress.Cypress['spec'] | null) {
-    this.spec = spec
+    if (this.isInitialBuildSucceed) {
```
Comment (lmiller1990) [inline packages/runner-ct/src/lib/state.ts:199]: nit: I think `hasInitialBuildSucceeded` is a little more natural.

File: packages/runner-ct/src/lib/event-manager.js [hunk 2]
```diff
@@ -195,6 +196,17 @@ const eventManager = {
       this._clearAllCookies()
     })
 
+    const hmrSocket = connectWebpackHmr({
+      url: `${window.location.origin}/cypress-webpack-hmr-socket`,
+      onReload: () => {
+        if (!state.isInitialBuildSucceed) {
```
Comment (lmiller1990) [inline packages/runner-ct/src/lib/event-manager.js:202]: not a blocker, but what do you think about putting the positive statement first in the if statement? eg:

```
if (state.initialBuildSucceed) {
  rerun() // most likely path
} else {
  // ... not often happening
}
```

just a style thing, no need to make this change to get the PR merged :)

File: packages/runner-ct/src/lib/webpack-hmr-client.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable no-console */
```
Comment (lmiller1990) [inline packages/runner-ct/src/lib/webpack-hmr-client.tsx:1]: probably does not need to be a tsx file.

File: packages/runner-ct/src/lib/webpack-hmr-client.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }
+
+  socketClient.onclose = (e) => {
+    console.error('Webpack HMR disconnected.')
+  }
+
+  return socketClient
+}
+
```
Comment (lmiller1990) [inline packages/runner-ct/src/lib/webpack-hmr-client.tsx:59]: Nice job figuring this fix out!

**Review by lmiller1990 — COMMENTED** [review]
Left some comments, none are really serious.

I also pulled the PR and tried it out. I clicked back and forth in tests for about 60 seconds, I was able to get a socket hang-up. I then tried again for about 5 minutes but did not get a socket hang up.

This was not updating source files, just clicking around. So at the very least it seems like a BIG improvement, maybe still some edge cases but we can continue to improve this as we go. Great!!

File: npm/webpack-dev-server/tsconfig.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     /* Basic Options */
-    "target": "es5" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', or 'ESNEXT'. */,
+    "target": "esnext" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', or 'ESNEXT'. */,
```
Comment (dmtrKovalenko) [inline npm/webpack-dev-server/tsconfig.json:4]: We don't need to transpile a lot for the node.js env

File: packages/runner-ct/src/lib/state.ts (hunk 1 above)
Comment (dmtrKovalenko) [reply packages/runner-ct/src/lib/state.ts:199]: I love the convention of all the boolean variables starting with `is` like previous `isRunning`, `isLoading` – it is much easier to understand the type without thinking :) 

Maybe you'll find better name starting with `is` for this variable? Thank you

**Comment by dmtrKovalenko** [comment]
@lmiller1990 thanks for testing it – I agree that it is not the 100% case especially because of [this setTimeout](https://github.com/cypress-io/cypress/blob/209b4837462e7f58d58956bce803f38694cfbbf1/packages/runner-ct/src/lib/state.ts#L207) – I really don't know why the even `ok` coming to us before files are served and ready to use.

File: packages/runner-ct/src/lib/state.ts (hunk 1 above)
Comment (elevatebart) [reply packages/runner-ct/src/lib/state.ts:199]: How about `isInitialBuildSuccesful` ? Not perfect temporality because it "was" succesful but I prefer it.

**Review by elevatebart — APPROVED** [review]
Great PR, 
My variable name suggestion could be discussed and taken after the PR is merged

Please merge this into feat/component-testing ;)

**Review by JessicaSachs — APPROVED** [review]
LGTM! The socket code is really great :+1:

File: packages/runner-ct/src/lib/event-manager.js [hunk 6]
```diff
@@ -195,6 +196,17 @@ const eventManager = {
...
+      url: `${window.location.origin}/cypress-webpack-hmr-socket`,
+      onReload: () => {
+        if (state.isInitialBuildSucceed) {
+          rerun()
+        } else {
+          state.initialBuildFired()
+        }
+      },
+    })
+
```
Comment (JessicaSachs) [inline packages/runner-ct/src/lib/event-manager.js:209]: Will need to refactor this to be webpack agnostic at some point. I added a ticket to Jira.

**Comment by dmtrKovalenko** [comment]
Closed in favor of #14488

**Pull request closed without merging**